lastuser_core/models/user.py

```diff
@@ -5,37 +5,116 @@
-from baseframe import _
+from baseframe import _, __
 
 from . import db, BaseMixin, UuidMixin
 
 
-__all__ = ['User', 'UserEmail', 'UserEmailClaim', 'PasswordResetRequest', 'UserExternalId',
+__all__ = ['Name', 'User', 'UserEmail', 'UserEmailClaim', 'PasswordResetRequest', 'UserExternalId',
            'UserPhone', 'UserPhoneClaim', 'Team', 'Organization', 'UserOldId', 'USER_STATUS']
 
 
-class USER_STATUS:
-    ACTIVE = 0     # Regular, active user
-    SUSPENDED = 1  # Suspended account
-    MERGED = 2     # Merged into another user
-    INVITED = 3    # Invited to make an account, doesn't have one yet
+class Name(UuidMixin, BaseMixin, db.Model):
+    """
+    Manage common namespace between the User and Organization models.
+    """
+    __tablename__ = 'name'
+    __uuid_primary_key__ = True
+
+    #: The "username" assigned to a user or organization (length limit 63 to fit DNS label limit)
+    name = db.Column(db.Unicode(63), nullable=False, unique=True)
+    # Only one of the following three may be set:
+    #: User that owns this name (limit one per user)
+    user_id = db.Column(None, db.ForeignKey('user.id', ondelete='CASCADE'), unique=True, nullable=True)
+    user = db.relationship('User', backref=db.backref('_name', uselist=False, cascade='all, delete-orphan'))
+    #: Organization that owns this name (limit one per organization)
+    org_id = db.Column(None, db.ForeignKey('organization.id', ondelete='CASCADE'), unique=True, nullable=True)
+    org = db.relationship('Organization', backref=db.backref('_name', uselist=False, cascade='all, delete-orphan'))
+    #: Reserved name (not assigned to any party)
+    reserved = db.Column(db.Boolean, nullable=False, default=False, index=True)
+
+    __table_args__ = (
+        db.CheckConstraint(
+            db.case([(user_id != None, 1)], else_=0) +
+            db.case([(org_id != None, 1)], else_=0) +
+            db.case([(reserved == True, 1)], else_=0) == 1,
+            name='username_owner_check'),  # NOQA
+        db.Index('ix_name_name_lower', db.func.lower(name).label('name_lower'),
+            unique=True, postgresql_ops={'name_lower': 'varchar_pattern_ops'})
+        )
+
+    @property
+    def owner(self):
+        return self.user or self.org
+
+    @owner.setter
+    def owner(self, value):
+        if isinstance(value, User):
+            self.user = value
+            self.org = None
+        elif isinstance(value, Organization):
+            self.user = None
+            self.org = value
+        else:
+            raise ValueError(value)
+        self.reserved = False
+
+    @classmethod
+    def get(cls, name):
+        return cls.query.filter(db.func.lower(Name.name) == db.func.lower(name)).one_or_none()
+
+    @db.validates('name')
+    def validate_name(self, key, value):
+        if not valid_username(value):
+            raise ValueError("Invalid username: " + value)
+        return value
+
 
+class SharedNameMixin(object):
```

Comment:
What is this model for?

Reply:
It's not a model. It offers the `name` property that makes the relationship behave like a string in both `User` and `Organization`.